File: 백준/Gold/2251. 물통/물통.java
import java.util.Scanner;
import java.util.ArrayList;
import java.util.Queue;
import java.util.LinkedList;
import java.util.Collections;

public class Main {
    static int A,B,C;
    static boolean[][][] visited = new boolean[201][201][201];
    
    public static void main(String args[]) {
      Scanner s = new Scanner(System.in);
      A = s.nextInt();
      B = s.nextInt();
      C = s.nextInt();
      ArrayList<int[]> list = new ArrayList<>();
      ArrayList<Integer> ans = new ArrayList<>();
      Queue<int[]> q = new LinkedList<>();
      
      q.add(new int[]{0,0,C});
      
      while(!q.isEmpty())
      {
          int[] temp = q.poll();
          
          if(visited[temp[0]][temp[1]][temp[2]]) continue;
          visited[temp[0]][temp[1]][temp[2]] = true;
          
          if (temp[0] == 0) ans.add(temp[2]);
		  
		  if (temp[0] + temp[1] > A) q.add(new int[] { A, temp[0] + temp[1] - A, temp[2] });
		  else q.add(new int[] { temp[0] + temp[1], 0, temp[2] });

		  if (temp[0] + temp[1] > B) q.add(new int[] { temp[0] + temp[1] - B, B, temp[2] });
          else q.add(new int[] { 0, temp[0] + temp[1], temp[2] });

		  if (temp[0] + temp[2] > A) q.add(new int[] { A, temp[1], temp[0] + temp[2] - A });
		  else q.add(new int[] { temp[0] + temp[2], temp[1], 0 });

		  if (temp[0] + temp[2] > C) q.add(new int[] { temp[0] + temp[2] - C, temp[1], C });
		  else q.add(new int[] { 0, temp[1], temp[0] + temp[2] });
		  
		  if (temp[1] + temp[2] > B) q.add(new int[] { temp[0], B, temp[1] + temp[2] - B });
		  else q.add(new int[] { temp[0], temp[1] + temp[2], 0 });

		  if (temp[1] + temp[2] > C) q.add(new int[] { temp[0], temp[1] + temp[2] - C, C });
		  else q.add(new int[] { temp[0], 0, temp[1] + temp[2] });
        }
        Collections.sort(ans);
        for(int i=0;i<ans.size();i++) System.out.print(ans.get(i) + " ");
    }
}

File: 백준/Gold/2251. 물통/물통.py
from collections import deque

visited = [[[0 for _ in range(201)] for _ in range(201)] for _ in range(201)]
A,B,C = map(int,input().split())
ans = []
q = deque([])
q.append([0,0,C])

while q:
    a,b,c = q.popleft()
    
    if visited[a][b][c]: continue
    visited[a][b][c] = True
    
    if a==0: ans.append(c)
    
    if a+b>B: q.append([a+b-B,B,c])
    else: q.append([0,a+b,c])
    
    if a + c > C: q.append([a+c-C,b,C])
    else: q.append([0,b,a+c])
    
    if b + a > A: q.append([A,b+a-A,c])
    else: q.append([b+a,0,c])
    
    if b + c > C: q.append([a,b+c-C,C])
    else: q.append([a,0,b+c])
    
    if c + a > A: q.append([A,b,c+a-A])
    else: q.append([c+a,b,0])
    
    if c + b > B: q.append([a,B,c+b-B])
    else: q.append([a,b+c,0])
    
ans.sort()
for i in range(len(ans)): print(ans[i],end=" ")